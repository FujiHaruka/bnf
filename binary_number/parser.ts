import {
  ScanContext,
  TokenNode,
  TokenType,
  UnexpectedTokenError,
} from "./core.ts";

export const TokenTypes = {
  "binary-number": new TokenType("binary-number"),
  "binary-sequence": new TokenType("binary-sequence"),
  "binary-digit": new TokenType("binary-digit"),
  "$literal": new TokenType("$literal"),
} as const;

type ParserAtom = (ctx: ScanContext) => TokenNode | null;

function expect(ctx: ScanContext, fn: ParserAtom): TokenNode {
  const nodeOrNull = fn(ctx);
  if (nodeOrNull) {
    return nodeOrNull;
  } else {
    throw new UnexpectedTokenError(ctx);
  }
}

export const Parser = {
  /**
   * `<binary-number> ::= <binary-digit> | "1" <binary-sequence>`
   */
  "binary-number"(ctx: ScanContext): TokenNode | null {
    const startAt = ctx.cursor;

    // if it starts with a specific characotr, it can match the latter case.
    if (ctx.currentChar === "1") {
      const literalNode = new TokenNode({
        type: TokenTypes.$literal,
        children: [],
        startAt: ctx.cursor,
        endAt: ctx.cursor + 1,
      });
      ctx.nextCursor();
      const node = Parser["binary-sequence"](ctx);
      if (node) {
        return new TokenNode({
          type: TokenTypes["binary-number"],
          children: [literalNode, node],
          startAt,
          endAt: startAt + 2,
        });
      }

      // Reset cursor position
      ctx.prevCursor();
    }

    const node = Parser["binary-digit"](ctx);
    if (node) {
      ctx.nextCursor();
      return new TokenNode({
        type: TokenTypes["binary-number"],
        children: [node],
        startAt: startAt,
        endAt: startAt + 1,
      });
    } else {
      return null;
    }
  },

  /**
   * `<binary-sequence> ::= <binary-digit> | <binary-digit> <binary-sequence>`
   */
  "binary-sequence"(ctx: ScanContext): TokenNode | null {
    const startAt = ctx.cursor;
    const children: TokenNode[] = [];

    let node: TokenNode | null = null;
    do {
      node = Parser["binary-digit"](ctx);
      if (node) {
        children.push(node);
      }
    } while (node);

    if (children.length === 0) {
      return null;
    } else {
      return new TokenNode({
        type: TokenTypes["binary-sequence"],
        children,
        startAt,
        endAt: startAt + children.length,
      });
    }
  },

  /**
   * `<binary-digit> ::= "0" | "1"`
   */
  "binary-digit"(ctx: ScanContext): TokenNode | null {
    const startAt = ctx.cursor;
    switch (ctx.currentChar) {
      case "0":
      case "1": {
        ctx.nextCursor();
        return new TokenNode({
          type: TokenTypes["binary-digit"],
          children: [],
          startAt,
          endAt: startAt + 1,
        });
      }
      default:
        return null;
    }
  },
};

export function parse(text: string): TokenNode {
  const ctx = new ScanContext(text);
  const node = expect(ctx, Parser["binary-number"]);
  if (!ctx.scanFinished) {
    throw new UnexpectedTokenError(ctx);
  }

  return node;
}
