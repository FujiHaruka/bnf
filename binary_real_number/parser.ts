import {
  LiteralTokenNode,
  longestNode,
  NamedTokenNode,
  TokenNode,
  TokenType,
  UnexpectedTokenError,
} from "./core.ts";

export type RuleName =
  | "binary-number"
  | "binary-decimal"
  | "binary-integer"
  | "binary-natural-number"
  | "binary-sequence"
  | "binary-digit"
  | "$literal";

export const TokenTypes: Readonly<Record<RuleName, TokenType>> = {
  "binary-number": new TokenType("binary-number"),
  "binary-decimal": new TokenType("binary-decimal"),
  "binary-integer": new TokenType("binary-integer"),
  "binary-natural-number": new TokenType("binary-natural-number"),
  "binary-sequence": new TokenType("binary-sequence"),
  "binary-digit": new TokenType("binary-digit"),
  "$literal": new TokenType("$literal"),
};

type ParserAtom<T extends TokenNode = TokenNode> = (
  text: string,
  position: number,
) => T | null;

function expect<T extends TokenNode>(
  fn: ParserAtom<T>,
): (text: string, startAt: number) => T {
  const fnOrThrow = (text: string, position: number) => {
    const nodeOrNull = fn(text, position);
    if (nodeOrNull) {
      return nodeOrNull;
    } else {
      throw new UnexpectedTokenError({
        char: text.charAt(position),
        position: position,
      });
    }
  };
  Object.defineProperty(fnOrThrow, "name", {
    value: `expect(${fn.name})`,
    configurable: true,
  });
  return fnOrThrow;
}

export const Parser = {
  /**
   * <binary-number> ::= <binary-integer> | <binary-decimal>
   */
  "binary-number"(text: string, position: number): NamedTokenNode | null {
    const candidates = [
      Parser["binary-integer"](text, position),
      Parser["binary-decimal"](text, position),
    ].filter((node: NamedTokenNode | null): node is NamedTokenNode =>
      Boolean(node)
    );
    const node = longestNode(candidates);
    if (node) {
      return new NamedTokenNode({
        type: TokenTypes["binary-number"],
        children: [node],
        startAt: position,
        endAt: node.endAt,
      });
    } else {
      return null;
    }
  },

  /**
   * <binary-decimal> ::= <binary-integer> "." <binary-sequence>
   */
  "binary-decimal"(text: string, position: number): NamedTokenNode | null {
    const integer = Parser["binary-integer"](text, position) ??
      Parser["binary-sequence"](text, position);
    if (integer) {
      const literal = expect(Parser["$literal"])(text, integer.endAt);
      if (literal.value !== ".") {
        throw new UnexpectedTokenError({
          char: literal.value,
          position: integer.endAt,
        });
      }

      const sequence = expect(Parser["binary-sequence"])(
        text,
        integer.endAt + 1,
      );
      return new NamedTokenNode({
        type: TokenTypes["binary-decimal"],
        children: [integer, literal, sequence],
        startAt: position,
        endAt: sequence.endAt,
      });
    } else {
      return null;
    }
  },

  /**
   * <binary-integer> ::= "0" | "-" <binary-natural-number> | <binary-natural-number>
   */
  "binary-integer"(text: string, position: number): NamedTokenNode | null {
    const char = text.charAt(position);
    switch (char) {
      case "0": {
        const literal = new LiteralTokenNode({
          type: TokenTypes["$literal"],
          value: char,
        });
        return new NamedTokenNode({
          type: TokenTypes["binary-integer"],
          children: [literal],
          startAt: position,
          endAt: position + 1,
        });
      }
      case "-": {
        const literal = new LiteralTokenNode({
          type: TokenTypes["$literal"],
          value: char,
        });
        const node = expect(Parser["binary-natural-number"])(
          text,
          position + 1,
        );
        return new NamedTokenNode({
          type: TokenTypes["binary-integer"],
          children: [literal],
          startAt: position,
          endAt: node.endAt,
        });
      }
      default: {
        const node = Parser["binary-natural-number"](text, position);
        if (node) {
          return new NamedTokenNode({
            type: TokenTypes["binary-integer"],
            children: [node],
            startAt: position,
            endAt: node.endAt,
          });
        } else {
          return null;
        }
      }
    }
  },

  /**
   * `<binary-natural-number> ::= <binary-digit> | "1" <binary-sequence>`
   */
  "binary-natural-number"(
    text: string,
    position: number,
  ): NamedTokenNode | null {
    // if it starts with a specific characotr, it can match the latter case.
    if (text.charAt(position) === "1") {
      const literal = expect(Parser["$literal"])(text, position);
      const sequence = Parser["binary-sequence"](text, position + 1);
      if (sequence) {
        return new NamedTokenNode({
          type: TokenTypes["binary-number"],
          children: [literal, sequence],
          startAt: position,
          endAt: sequence.endAt,
        });
      }
    }

    const digit = Parser["binary-digit"](text, position);
    if (digit) {
      return new NamedTokenNode({
        type: TokenTypes["binary-number"],
        children: [digit],
        startAt: position,
        endAt: digit.endAt,
      });
    } else {
      return null;
    }
  },

  /**
   * `<binary-sequence> ::= <binary-digit> | <binary-digit> <binary-sequence>`
   */
  "binary-sequence"(text: string, position: number): NamedTokenNode | null {
    const children: NamedTokenNode[] = [];

    let node: NamedTokenNode | null = null;
    let endAt = position;
    do {
      node = Parser["binary-digit"](text, endAt);
      if (node) {
        endAt = node.endAt;
        children.push(node);
      }
    } while (node);

    if (children.length === 0) {
      return null;
    } else {
      return new NamedTokenNode({
        type: TokenTypes["binary-sequence"],
        children,
        startAt: position,
        endAt,
      });
    }
  },

  /**
   * `<binary-digit> ::= "0" | "1"`
   */
  "binary-digit"(text: string, position: number): NamedTokenNode | null {
    switch (text.charAt(position)) {
      case "0":
      case "1": {
        const node = expect(Parser["$literal"])(text, position);
        return new NamedTokenNode({
          type: TokenTypes["binary-digit"],
          children: [node],
          startAt: position,
          endAt: position + 1,
        });
      }
      default:
        return null;
    }
  },

  "$literal"(text: string, position: number): LiteralTokenNode | null {
    return new LiteralTokenNode({
      type: TokenTypes["$literal"],
      value: text.charAt(position),
    });
  },
};

export function parse(text: string): NamedTokenNode {
  const node = expect(Parser["binary-number"])(text, 0);
  if (node.endAt !== text.length) {
    throw new UnexpectedTokenError({
      char: text.charAt(node.endAt),
      position: node.endAt,
    });
  }

  return node;
}
