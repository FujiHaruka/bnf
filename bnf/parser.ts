import { cleanupTempTokenNodes } from "./combinators/cleanupTempTokenNodes.ts";
import { concat } from "./combinators/concat.ts";
import { flattenRecursiveNodes } from "./combinators/flattenRecursiveNodes.ts";
import { literalTokenParser } from "./combinators/literalTokenParser.ts";
import { or } from "./combinators/or.ts";
import { NamedTokenNode, TempTokenType, TokenType } from "./token.ts";
import { UnexpectedTokenError } from "./utils/errors.ts";
import { Result } from "./utils/Result.ts";

export const RuleNames = [
  "syntax",
  "rule",
  "opt-whitespace",
  "expression",
  "line-end",
  "list",
  "term",
  "literal",
  "text1",
  "text2",
  "character",
  "letter",
  "digit",
  "symbol",
  "character1",
  "character2",
  "rule-name",
  "rule-char",
] as const;

export type RuleName = typeof RuleNames[number];

export const TokenTypes = Object.fromEntries(RuleNames
  .map((ruleName) => [ruleName, new TokenType(ruleName)])) as Readonly<
    Record<RuleName, TokenType>
  >;

export const Parser = {
  /**
   * <syntax> ::= <rule> | <rule> <syntax>
   */
  "syntax"(text: string, position: number): Result<NamedTokenNode> {
    return or(TokenTypes["syntax"], [
      Parser["rule"],
      concat(TempTokenType, [
        Parser["rule"],
        Parser["syntax"],
      ]),
    ])(text, position);
  },
  /**
   * <rule> ::= <opt-whitespace> "<" <rule-name> ">" <opt-whitespace> "::=" <opt-whitespace> <expression> <line-end>
   */
  "rule"(text: string, position: number): Result<NamedTokenNode> {
    return concat(TokenTypes["rule"], [
      Parser["opt-whitespace"],
      literalTokenParser("<"),
      Parser["rule-name"],
      literalTokenParser(">"),
      Parser["opt-whitespace"],
      // TODO: how to express "::="?
      literalTokenParser(":"),
      literalTokenParser(":"),
      literalTokenParser("="),
      Parser["opt-whitespace"],
      Parser["expression"],
      Parser["line-end"],
    ])(text, position);
  },
  /**
   * <opt-whitespace> ::= " " <opt-whitespace> | ""
   */
  "opt-whitespace"(text: string, position: number): Result<NamedTokenNode> {
    return or(TokenTypes["opt-whitespace"], [
      concat(TempTokenType, [
        literalTokenParser(" "),
        Parser["opt-whitespace"],
      ]),
      // TODO: empty string parser?
      literalTokenParser(""),
    ])(text, position);
  },
  /**
   * <expression>     ::= <list> | <list> <opt-whitespace> "|" <opt-whitespace> <expression>
   */
  "expression"(text: string, position: number): Result<NamedTokenNode> {
    return or(TokenTypes["expression"], [
      Parser["list"],
      concat(TempTokenType, [
        Parser["list"],
        Parser["opt-whitespace"],
        literalTokenParser("|"),
        Parser["opt-whitespace"],
        Parser["expression"],
      ]),
    ])(text, position);
  },
  /**
   * <line-end> ::= <opt-whitespace> <EOL> | <line-end> <line-end>
   */
  "line-end"(text: string, position: number): Result<NamedTokenNode> {
    return or(TokenTypes["line-end"], [
      concat(TempTokenType, [
        Parser["opt-whitespace"],
        // TODO: EOL parser
        literalTokenParser("\n"),
      ]),
      concat(TempTokenType, [
        Parser["line-end"],
        Parser["line-end"],
      ]),
    ])(text, position);
  },
  /**
   * <list> ::= <term> | <term> <opt-whitespace> <list>
   */
  "list"(text: string, position: number): Result<NamedTokenNode> {
    return or(TokenTypes["list"], [
      Parser["term"],
      concat(TempTokenType, [
        Parser["term"],
        Parser["opt-whitespace"],
        Parser["list"],
      ]),
    ])(text, position);
  },
  /**
   * <term> ::= <literal> | "<" <rule-name> ">"
   */
  "term"(text: string, position: number): Result<NamedTokenNode> {
    return or(TokenTypes["term"], [
      Parser["literal"],
      concat(TempTokenType, [
        literalTokenParser("<"),
        Parser["rule-name"],
        literalTokenParser(">"),
      ]),
    ])(text, position);
  },
  /**
   * <literal> ::= '"' <text1> '"' | "'" <text2> "'"
   */
  "literal"(text: string, position: number): Result<NamedTokenNode> {
    return or(TokenTypes["literal"], [
      concat(TempTokenType, [
        literalTokenParser('"'),
        Parser["text1"],
        literalTokenParser('"'),
      ]),
      concat(TempTokenType, [
        literalTokenParser("'"),
        Parser["text2"],
        literalTokenParser("'"),
      ]),
    ])(text, position);
  },
  /**
   * <text1> ::= "" | <character1> <text1>
   */
  "text1"(text: string, position: number): Result<NamedTokenNode> {
    return or(TokenTypes["text2"], [
      Parser["character1"],
      Parser["text1"],
    ])(text, position);
  },
  /**
   * <text2> ::= '' | <character2> <text2>
   */
  "text2"(text: string, position: number): Result<NamedTokenNode> {
    return or(TokenTypes["text2"], [
      Parser["character2"],
      Parser["text2"],
    ])(text, position);
  },
  /**
   * <character> ::= <letter> | <digit> | <symbol>
   */
  "character"(text: string, position: number): Result<NamedTokenNode> {
    return or(TokenTypes["character"], [
      Parser["letter"],
      Parser["digit"],
      Parser["symbol"],
    ])(text, position);
  },

  /**
   * <letter> ::= "A" | "B" | "C" | "D" | "E" | "F" | "G" | "H" | "I" | "J" | "K" | "L" | "M" | "N" | "O" | "P" | "Q" | "R" | "S" | "T" | "U" | "V" | "W" | "X" | "Y" | "Z" | "a" | "b" | "c" | "d" | "e" | "f" | "g" | "h" | "i" | "j" | "k" | "l" | "m" | "n" | "o" | "p" | "q" | "r" | "s" | "t" | "u" | "v" | "w" | "x" | "y" | "z"
   */
  "letter"(text: string, position: number): Result<NamedTokenNode> {
    return or(
      TokenTypes.letter,
      [
        "A",
        "B",
        "C",
        "D",
        "E",
        "F",
        "G",
        "H",
        "I",
        "J",
        "K",
        "L",
        "M",
        "N",
        "O",
        "P",
        "Q",
        "R",
        "S",
        "T",
        "U",
        "V",
        "W",
        "X",
        "Y",
        "Z",
        "a",
        "b",
        "c",
        "d",
        "e",
        "f",
        "g",
        "h",
        "i",
        "j",
        "k",
        "l",
        "m",
        "n",
        "o",
        "p",
        "q",
        "r",
        "s",
        "t",
        "u",
        "v",
        "w",
        "x",
        "y",
        "z",
      ].map((char) => literalTokenParser(char)),
    )(text, position);
  },
  /**
   * <digit> ::= "0" | "1" | "2" | "3" | "4" | "5" | "6" | "7" | "8" | "9"
   */
  "digit"(text: string, position: number): Result<NamedTokenNode> {
    return or(
      TokenTypes.letter,
      [
        "0",
        "1",
        "2",
        "3",
        "4",
        "5",
        "6",
        "7",
        "8",
        "9",
      ].map((char) => literalTokenParser(char)),
    )(text, position);
  },
  /**
   * <symbol> ::=  "|" | " " | "!" | "#" | "$" | "%" | "&" | "(" | ")" | "*" | "+" | "," | "-" | "." | "/" | ":" | ";" | ">" | "=" | "<" | "?" | "@" | "[" | "\" | "]" | "^" | "_" | "`" | "{" | "}" | "~"
   */
  "symbol"(text: string, position: number): Result<NamedTokenNode> {
    return or(
      TokenTypes.letter,
      [
        "|",
        " ",
        "!",
        "#",
        "$",
        "%",
        "&",
        "(",
        ")",
        "*",
        "+",
        ",",
        "-",
        ".",
        "/",
        ":",
        ";",
        ">",
        "=",
        "<",
        "?",
        "@",
        "[",
        "\\",
        "]",
        "^",
        "_",
        "`",
        "{",
        "}",
        "~",
      ].map((char) => literalTokenParser(char)),
    )(text, position);
  },
  /**
   * <character1> ::= <character> | "'"
   */
  "character1"(text: string, position: number): Result<NamedTokenNode> {
    return or(TokenTypes["character1"], [
      Parser["character"],
      literalTokenParser("'"),
    ])(text, position);
  },
  /**
   * <character2> ::= <character> | '"'
   */
  "character2"(text: string, position: number): Result<NamedTokenNode> {
    return or(TokenTypes["character2"], [
      Parser["character"],
      literalTokenParser('"'),
    ])(text, position);
  },
  /**
   * <rule-name> ::= <letter> | <rule-name> <rule-char>
   */
  "rule-name"(text: string, position: number): Result<NamedTokenNode> {
    return or(TokenTypes["rule-name"], [
      Parser["letter"],
      concat(TempTokenType, [
        Parser["rule-name"],
        Parser["rule-char"],
      ]),
    ])(text, position);
  },
  /**
   * <rule-char> ::= <letter> | <digit> | "-"
   */
  "rule-char"(text: string, position: number): Result<NamedTokenNode> {
    return or(TokenTypes["rule-char"], [
      Parser["letter"],
      Parser["digit"],
      literalTokenParser("-"),
    ])(text, position);
  },
};

export function parse(text: string): NamedTokenNode {
  const node = Parser["syntax"](text, 0)
    .map<NamedTokenNode>(cleanupTempTokenNodes)
    .map<NamedTokenNode>(flattenRecursiveNodes)
    .unwrap();

  if (node.endAt !== text.length) {
    throw new UnexpectedTokenError({
      ruleName: "$root",
      char: text.charAt(node.endAt),
      position: node.endAt,
    });
  }

  return node;
}
