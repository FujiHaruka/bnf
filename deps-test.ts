export {
  afterEach,
  beforeEach,
  describe,
  it,
} from "https://deno.land/std@0.158.0/testing/bdd.ts";
export {
  assert,
  assertEquals,
  assertInstanceOf,
  assertIsError,
  assertObjectMatch,
  assertThrows,
} from "https://deno.land/std@0.158.0/testing/asserts.ts";
export { assertSnapshot } from "https://deno.land/std@0.158.0/testing/snapshot.ts";
export * as posix from "https://deno.land/std@0.158.0/path/posix.ts";
import makeloc from "https://deno.land/x/dirname@1.1.2/mod.ts";
export { makeloc };
