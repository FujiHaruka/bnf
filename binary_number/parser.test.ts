import { assertSnapshot, assertThrows } from "../deps-test.ts";
import { parse } from "./parser.ts";

function toJsonObject(obj: unknown) {
  return JSON.parse(JSON.stringify(obj));
}

Deno.test("parse / success", async (t) => {
  const tests: string[] = [
    "0",
    "1",
    "10",
    "11",
    "100",
    "101",
    "110",
    "111",
  ];

  for (const test of tests) {
    await t.step(`"${test}"`, async (t) => {
      console.log(JSON.stringify(toJsonObject(parse(test))));
      await assertSnapshot(t, toJsonObject(parse(test)));
    });
  }
});

Deno.test("parse / failure", async (t) => {
  const tests: string[] = [
    "",
    "a",
    "01",
    "10a",
  ];

  for (const test of tests) {
    await t.step(`"${test}"`, async () => {
      await assertThrows(() => parse(test));
    });
  }
});
