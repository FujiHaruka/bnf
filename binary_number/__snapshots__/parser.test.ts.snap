export const snapshot = {};

snapshot[`parse / success > "0" 1`] = `
{
  children: [
    {
      children: [
      ],
      endAt: 1,
      startAt: 0,
      type: "binary-digit",
    },
  ],
  endAt: 1,
  startAt: 0,
  type: "binary-number",
}
`;

snapshot[`parse / success > "10" 1`] = `
{
  children: [
    {
      children: [
      ],
      endAt: 1,
      startAt: 0,
      type: "\$literal",
    },
    {
      children: [
        {
          children: [
          ],
          endAt: 2,
          startAt: 1,
          type: "binary-digit",
        },
      ],
      endAt: 2,
      startAt: 1,
      type: "binary-sequence",
    },
  ],
  endAt: 2,
  startAt: 0,
  type: "binary-number",
}
`;

snapshot[`parse / success > "1" 1`] = `
{
  children: [
    {
      children: [
      ],
      endAt: 1,
      startAt: 0,
      type: "binary-digit",
    },
  ],
  endAt: 1,
  startAt: 0,
  type: "binary-number",
}
`;

snapshot[`parse / success > "11" 1`] = `
{
  children: [
    {
      children: [
      ],
      endAt: 1,
      startAt: 0,
      type: "\$literal",
    },
    {
      children: [
        {
          children: [
          ],
          endAt: 2,
          startAt: 1,
          type: "binary-digit",
        },
      ],
      endAt: 2,
      startAt: 1,
      type: "binary-sequence",
    },
  ],
  endAt: 2,
  startAt: 0,
  type: "binary-number",
}
`;

snapshot[`parse / success > "100" 1`] = `
{
  children: [
    {
      children: [
      ],
      endAt: 1,
      startAt: 0,
      type: "\$literal",
    },
    {
      children: [
        {
          children: [
          ],
          endAt: 2,
          startAt: 1,
          type: "binary-digit",
        },
        {
          children: [
          ],
          endAt: 3,
          startAt: 2,
          type: "binary-digit",
        },
      ],
      endAt: 3,
      startAt: 1,
      type: "binary-sequence",
    },
  ],
  endAt: 2,
  startAt: 0,
  type: "binary-number",
}
`;

snapshot[`parse / success > "101" 1`] = `
{
  children: [
    {
      children: [
      ],
      endAt: 1,
      startAt: 0,
      type: "\$literal",
    },
    {
      children: [
        {
          children: [
          ],
          endAt: 2,
          startAt: 1,
          type: "binary-digit",
        },
        {
          children: [
          ],
          endAt: 3,
          startAt: 2,
          type: "binary-digit",
        },
      ],
      endAt: 3,
      startAt: 1,
      type: "binary-sequence",
    },
  ],
  endAt: 2,
  startAt: 0,
  type: "binary-number",
}
`;

snapshot[`parse / success > "110" 1`] = `
{
  children: [
    {
      children: [
      ],
      endAt: 1,
      startAt: 0,
      type: "\$literal",
    },
    {
      children: [
        {
          children: [
          ],
          endAt: 2,
          startAt: 1,
          type: "binary-digit",
        },
        {
          children: [
          ],
          endAt: 3,
          startAt: 2,
          type: "binary-digit",
        },
      ],
      endAt: 3,
      startAt: 1,
      type: "binary-sequence",
    },
  ],
  endAt: 2,
  startAt: 0,
  type: "binary-number",
}
`;

snapshot[`parse / success > "111" 1`] = `
{
  children: [
    {
      children: [
      ],
      endAt: 1,
      startAt: 0,
      type: "\$literal",
    },
    {
      children: [
        {
          children: [
          ],
          endAt: 2,
          startAt: 1,
          type: "binary-digit",
        },
        {
          children: [
          ],
          endAt: 3,
          startAt: 2,
          type: "binary-digit",
        },
      ],
      endAt: 3,
      startAt: 1,
      type: "binary-sequence",
    },
  ],
  endAt: 2,
  startAt: 0,
  type: "binary-number",
}
`;
